language/resource-api/README.md

```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
 > values changed
 ```
 
+### Provider feature: `custom_insync`
+
+Allows for overriding default property `insync?` checking using custom logic in the provider.
+
+```ruby
+# lib/puppet/type/test_custom_insync.rb
+Puppet::ResourceApi.register_type(
+  name: 'test_custom_insync',
+  docs: <<-EOS,
+      This type provides Puppet with the capabilities to manage ...
+    EOS
+  features: ['custom_insync'],
+  attributes:   {
+    ensure:      {
+      type:    'Enum[present, absent]',
+      desc:    'Whether this resource should be present or absent on the target system.',
+      default: 'present',
+    },
+    name:        {
+      type:      'String',
+      desc:      'The name of the resource you want to manage.',
+      behaviour: :namevar,
+    },
+    some_array:  {
+      type: 'Optional[Array[String]]',
+      desc: 'An array aiding array attestation',
+    },
+    force: {
+      type:      'Boolean',
+      desc:      'If true, the specified array is inclusive and must match the existing state exactly.',
+      behaviour: :parameter,
+      default:   false,
+    },
+  },
+)
+
+# lib/puppet/provider/test_custom_insync/test_custom_insync.rb
+class Puppet::Provider::TestCustomInsync::TestCustomInsync < Puppet::ResourceApi::SimpleProvider
+  def insync?(context, name, property_name, is_hash, should_hash)
+    context.debug("Checking whether #{property_name} is out of sync")
+    case property_name
+    when :some_array
+      if should_hash[:force]
+        context.debug("Checking an order independent array")
+        return is_hash[property_name].sort == should_hash[property_name].sort
+      else
+        context.debug("Checking a subset match array")
+        found_members = (is_hash[property_name] & should_hash[property_name]).sort
+        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
+        return missing_members.empty? ? true : "Adding missing members #{missing_members}"
+      end
+    end
+end
+```
+
+In normal operations, the runtime environment uses strict value equality to determine if a resource instance needs to be state-enforced.
+When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
+This can be used to implement custom state comparisons, when the default value-equality is not sufficient.
+
+The `insync?` method is executed exactly once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
+If no properties were specified in the instance of a resource, `insync?` is not called.
+The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.
```

Comment:
Is this desired state?

Reply:
It determines if the resource is _in_ the desired state.

Reply:
Maybe rephrase it like this?